### PR TITLE
Support retrieving workspace of path dependencies in cargo

### LINKED ISSUE #9533: Rust: `failed to find a workspace root` for workspace dependency in a non-workspace crate
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package ecosystem

cargo

### Package manager version

1.77

### Language version

1.77

### Manifest location and content before the Dependabot update

_No response_

### dependabot.yml content

https://github.com/apache/arrow-datafusion/blob/b2f5e39c4636951ccb7357478f21fcf34a692f51/.github/dependabot.yml

### Updated dependency

_No response_

### What you expected to see, versus what you actually saw

We have a repository with a workspace (datafusion):

https://github.com/apache/arrow-datafusion/blob/b2f5e39c4636951ccb7357478f21fcf34a692f51/Cargo.toml#L18-L43

However we have a nested crate which is excluded from the workspace:

https://github.com/apache/arrow-datafusion/blob/b2f5e39c4636951ccb7357478f21fcf34a692f51/datafusion-cli/Cargo.toml

This nested crate (datafusion-cli) has a dependency on datafusion. We are getting this error when trying to setup dependabot for datafusion-cli:

```
 Dependabot can't resolve your Rust dependency files

Dependabot failed to update your dependencies because there was an error resolving your Rust dependency files.

Dependabot encountered the following error:

Updating crates.io index
error: failed to get `datafusion` as a dependency of package `datafusion-cli v37.0.0 (dependabot_tmp_dir/datafusion-cli)`

Caused by:
  failed to load source for dependency `datafusion`

Caused by:
  Unable to update dependabot_tmp_dir/datafusion/core

Caused by:
  failed to parse manifest at `dependabot_tmp_dir/datafusion/core/Cargo.toml`

Caused by:
  error inheriting `version` from workspace root manifest's `workspace.package.version`

Caused by:
  failed to find a workspace root
```

Usually to build/check/run datafusion-cli we would `cd` into the `datafusion-cli/` dir and run the cargo commands there, and that would work to build and resolve dependencies just fine.

### Native package manager behavior

_No response_

### Images of the diff or a link to the PR, issue, or logs

_No response_

### Smallest manifest that reproduces the issue

_No response_

### PR DESCRIPTION
**Pull request by Jefffrey**
### What are you trying to accomplish?

Given a repository like so:

```
.
├── cli_crate
│   ├── Cargo.toml
└── workspace
    ├── Cargo.toml
    └── nested
        └── Cargo.toml
```

Where `workspace` represents a workspace containing only `nested`, and `cli_crate` has a path dependency to `nested`, and `nested` inherits any property from the root (e.g. `edition =  { workspace = true }`).

If we point dependabot to only `cli_crate`, it will error out when running `cargo update` as though it can fetch the path dependency `nested`, it cannot resolve this as it also needs the `workspace` due to this inheritance. Fix Cargo file_fetcher to support this logic of locating the root of a workspace member who is a path dependency, and fetching it.

For negative cases, where the root cannot be located or is incorrect, emit an error message (instead of when running `cargo update` which can be confusing to understand).

Closes #9533

### How will you know you've accomplished your goal?

Unit tests added (see reference repository: https://github.com/Jefffrey/dependabot-cargo-test/tree/master)

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

File: cargo/spec/dependabot/cargo/file_fetcher_spec.rb [hunk 1]
```diff
@@ -678,7 +678,7 @@
 
       stub_request(:get, url + "excluded/Cargo.toml?ref=sha")
         .with(headers: { "Authorization" => "token token" })
-        .to_return(status: 200, body: member_fixture, headers: json_header)
+        .to_return(status: 200, body: excluded_fixture, headers: json_header)
```
Comment (Jefffrey) [inline cargo/spec/dependabot/cargo/file_fetcher_spec.rb:681]: Small typo (I assume?) I found when looking at https://github.com/dependabot/dependabot-core/pull/7649 for reference

**Review by abdulapopoola — APPROVED** [review]
Approving based on @RobJellinghaus ' approval :)

**Comment by RobJellinghaus** [comment]
> Approving based on @RobJellinghaus ' approval :)

Thanks Abdul! Sorry I don't have approver ability myself for some reason, let me know if you can diagnose that. Thanks @Jefffrey for this work! Do you have any other Rust Dependabot issues you are looking at?

**Comment by abdulapopoola** [comment]
Thanks so much @Jefffrey !

**Comment by mistydemeo** [comment]
This led to a regression that causes Dependabot runs to fail in multi-package Cargo workspaces - see #10584. Any chance this could either get reverted or fixed forward?

**Comment by Jefffrey** [comment]
> This led to a regression that causes Dependabot runs to fail in multi-package Cargo workspaces - see #10584. Any chance this could either get reverted or fixed forward?

Sorry about this! I've raised a PR to revert these changes: #10599

**Pull request merged**